src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs

```diff
@@ -0,0 +1,288 @@
+
+            var groupedSpans = classifiedSpans.GroupBy(s => s.TextSpan);
+
+            using var _ = ArrayBuilder<int>.GetInstance(out var data);
+
+            // TO-DO: We should implement support for streaming once this LSP bug is fixed:
+            // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1132601
+            var tokens = ComputeTokens(
+                data, ref lastLineNumber, ref lastStartCharacter, text.Lines, groupedSpans);
+
+            return new LSP.SemanticTokens { ResultId = (previousResultId + 1).ToString(), Data = tokens };
+        }
+
+        private static int[] ComputeTokens(
+            ArrayBuilder<int> data,
+            ref int lastLineNumber,
+            ref int lastStartCharacter,
+            TextLineCollection lines,
+            IEnumerable<IGrouping<TextSpan, ClassifiedSpan>> groupedSpans)
+        {
+            foreach (var span in groupedSpans)
+            {
+                ComputeNextToken(data, lines, ref lastLineNumber, ref lastStartCharacter, span);
+            }
+
+            return data.ToArray();
+        }
+
+        private static void ComputeNextToken(
+            ArrayBuilder<int> data,
+            TextLineCollection lines,
+            ref int lastLineNumber,
+            ref int lastStartCharacter,
+            IGrouping<TextSpan, ClassifiedSpan> span)
+        {
+            // Each semantic token is represented in LSP by five numbers:
+            //     1. Token line number, relative to the previous token
+            //     2. Token start character, relative to the previous token
+            //     3. Token length
+            //     4. Token type (index) - looked up in SemanticTokensLegend.tokenTypes
+            //     5. Token modifiers - each set bit will be looked up in SemanticTokensLegend.tokenModifiers
+
+            var textSpan = span.Key;
+            var linePosition = lines.GetLinePositionSpan(textSpan).Start;
+            var lineNumber = linePosition.Line;
+            var startCharacter = linePosition.Character;
+
+            // 1. Token line number, relative to the previous token
+            var deltaLine = lineNumber - lastLineNumber;
+            data.Add(deltaLine);
+
+            // 2. Token start character, relative to the previous token
+            // (Relative to 0 or the previous token’s start if they're on the same line)
+            if (lastLineNumber == lineNumber)
+            {
+                data.Add(startCharacter - lastStartCharacter);
+            }
+            else
+            {
+                data.Add(startCharacter);
+            }
+
+            // 3. Token length
+            data.Add(textSpan.Length);
+
+            var additiveResults = span.Where(s => ClassificationTypeNames.AdditiveTypeNames.Contains(s.ClassificationType));
+
+            // 4. Token type - looked up in SemanticTokensLegend.tokenTypes.
+            var tokenTypeClassifiedSpan = span.Except(additiveResults).Single();
```

Comment:
all this linq seems super expensive.  we optimized the heck out of classification to ensure low allocations and lightweight types.  it feels like we should be able to preserve this with a straight linear pass over the results.

Reply:
The behavior here is modeled after Omnisharp: https://github.com/OmniSharp/omnisharp-roslyn/blob/master/src/OmniSharp.Roslyn.CSharp/Services/SemanticHighlight/SemanticHighlightService.cs#L53-L65

I can definitely see the concern with the LINQ queries. Currently ClassificationTypeNames.AdditiveTypeNames only contains one value, so it's relatively inexpensive, but I see how that can change in the future.

If you would be available to chat about this sometime, please let me know. I can definitely brainstorm more tomorrow, but I can't currently think of a way to do this with a straight linear pass (although my brain is a bit foggy rn).